internal/wasm/module.go

```diff
@@ -895,3 +933,66 @@ func ExternTypeName(et ExternType) string {
 	}
 	return fmt.Sprintf("%#x", et)
 }
+
+// A FunctionListener can be registered for any function via FunctionListenerFactory to
+// be notified when the function is called.
+type FunctionListener interface {
+	// Before is invoked before a function is called. ctx is the context of the caller function.
+	// The returned context will be used as the context of this function call. To add context
+	// information for this function call, add it to ctx and return the updated context. If
+	// no context informationis needed, return ctx as is.
+	Before(ctx context.Context) context.Context
```

Comment:
For now values aren't passed here. I don't know how useful values would be in general given they're usually some opaque ABI-related value or pointer.

Reply:
maybe extract this sort of thing to listener.go/listener_test.go